File: apps/api/src/services/workout_categories.py
from __future__ import annotations

from dataclasses import dataclass

DAY_CATEGORY_CYCLE = "cycle"
DAY_CATEGORY_WEIGHTS = "weights"
DAY_CATEGORY_FLEXIBILITY = "flexibility"
DAY_CATEGORY_WALK = "walk"
DAY_CATEGORY_REST = "rest"

#: The workout-type vocabulary, mirroring ``packages/shared/src/workoutTypes.ts``
#: entry for entry (Batch 253, CR236-05). ``PlannedWorkout.workout_type`` is
#: unconstrained free text classified by ten separate pieces of code — nine here
#: and one in TypeScript — and the two languages implemented **different rules**:
#: sets plus prefixes here, regexes there. They disagreed for values the app's own
#: label map lists. ``flexibility`` and ``deliberate_walk`` classified as
#: ``flexibility``/``walk`` in the app and ``weights`` here; a bare ``vo2`` or
#: ``endurance`` classified as ``cycle`` there and ``weights`` here. Latent, because
#: production holds seven values that classify identically — and live the moment
#: the block generator, the quick-add sheet or a hand-authored plan writes an
#: eighth, with the app and the coach then disagreeing about what kind of day it is
#: and nothing erroring anywhere.
WORKOUT_TYPE_CATEGORY: dict[str, str] = {
    "bike_z2": DAY_CATEGORY_CYCLE,
    "bike_endurance": DAY_CATEGORY_CYCLE,
    "bike_recovery": DAY_CATEGORY_CYCLE,
    "bike_tempo": DAY_CATEGORY_CYCLE,
    "bike_sweet_spot": DAY_CATEGORY_CYCLE,
    "bike_threshold": DAY_CATEGORY_CYCLE,
    "bike_vo2": DAY_CATEGORY_CYCLE,
    # Bare discipline words. TypeScript already classified these as ``cycle`` by
    # regex while this side's set lookup sent them to ``weights``; naming them
    # resolves the divergence the way a human reads them rather than the way the
    # stricter side happened to fall.
    "z2": DAY_CATEGORY_CYCLE,
    "endurance": DAY_CATEGORY_CYCLE,
    "recovery_ride": DAY_CATEGORY_CYCLE,
    "tempo": DAY_CATEGORY_CYCLE,
    "sweet_spot": DAY_CATEGORY_CYCLE,
    "threshold": DAY_CATEGORY_CYCLE,
    "vo2": DAY_CATEGORY_CYCLE,
    "strength": DAY_CATEGORY_WEIGHTS,
    "strength_maintenance": DAY_CATEGORY_WEIGHTS,
    "strength_recovery": DAY_CATEGORY_WEIGHTS,
    "mobility": DAY_CATEGORY_FLEXIBILITY,
    "flexibility": DAY_CATEGORY_FLEXIBILITY,
    "walk": DAY_CATEGORY_WALK,
    "walking": DAY_CATEGORY_WALK,
    "walk_recovery": DAY_CATEGORY_WALK,
    "deliberate_walk": DAY_CATEGORY_WALK,
}

WORKOUT_TYPES: frozenset[str] = frozenset(WORKOUT_TYPE_CATEGORY)

#: Retained because callers read them; derived from the one table so they cannot
#: drift from it.
WORKOUT_TYPE_CYCLE = {k for k, v in WORKOUT_TYPE_CATEGORY.items() if v == DAY_CATEGORY_CYCLE}
WORKOUT_TYPE_WEIGHTS = {k for k, v in WORKOUT_TYPE_CATEGORY.items() if v == DAY_CATEGORY_WEIGHTS}
WORKOUT_TYPE_FLEXIBILITY = {
    k for k, v in WORKOUT_TYPE_CATEGORY.items() if v == DAY_CATEGORY_FLEXIBILITY
}
WORKOUT_TYPE_WALK = {k for k, v in WORKOUT_TYPE_CATEGORY.items() if v == DAY_CATEGORY_WALK}


@dataclass(frozen=True)
class DayState:
    categories: list[str]
    label: str
    is_rest: bool


def category_for_workout_type(workout_type: str | None) -> str:
    """The one Python classifier. Mirrors ``categoryForWorkoutType`` exactly.

    The three prefix fallbacks are deliberate and are the same three TypeScript
    applies, so a ``bike_something_new`` reaching the column before the vocabulary
    catches up classifies identically on both sides rather than differently.
    Anything else is ``weights`` — the historical default, kept as the default on
    both sides so the two cannot diverge on the unknown case either.
    """
    value = (workout_type or "").strip().lower()
    known = WORKOUT_TYPE_CATEGORY.get(value)
    if known is not None:
        return known
    if value.startswith("bike_"):
        return DAY_CATEGORY_CYCLE
    if value.startswith("strength_"):
        return DAY_CATEGORY_WEIGHTS
    if value.startswith("walk_"):
        return DAY_CATEGORY_WALK
    return DAY_CATEGORY_WEIGHTS


def is_known_workout_type(value: str) -> bool:
    return value.strip().lower() in WORKOUT_TYPE_CATEGORY


def day_state_for_workout_types(workout_types: list[str]) -> DayState:
    categories: list[str] = []
    for workout_type in workout_types:
        category = category_for_workout_type(workout_type)
        if category not in categories:
            categories.append(category)
    if not categories:
        return DayState(categories=[DAY_CATEGORY_REST], label="Rest", is_rest=True)
    labels = {
        DAY_CATEGORY_CYCLE: "Cycle",
        DAY_CATEGORY_WEIGHTS: "Weights",
        DAY_CATEGORY_FLEXIBILITY: "Flexibility",
        DAY_CATEGORY_WALK: "Walk",
    }
    return DayState(
        categories=categories,
        label=" + ".join(labels[category] for category in categories),
        is_rest=False,
    )


def is_bike_workout_type(workout_type: str | None) -> bool:
    return category_for_workout_type(workout_type) == DAY_CATEGORY_CYCLE


def normalise_workout_type(value: str | None) -> str:
    """Coerce a written ``workout_type`` into the shared vocabulary.

    Batch 253 (CR236-05). Three writers put unvalidated text into this column, and
    one of them is the **model**: ``block_generator`` writes
    ``str(workout["workoutType"])`` straight from generated JSON. A value outside
    the vocabulary does not error anywhere — it silently classifies, possibly
    differently in each language, and the app and the coach then disagree about
    what kind of day it is.

    Normalising rather than rejecting is deliberate. A ``CHECK`` constraint was
    considered and **not** taken: its failure mode is a 500 during plan generation
    on a column ten call sites already read defensively, which is worse than the
    misclassification it would prevent. This coerces case and whitespace, keeps any
    value the vocabulary or its three prefixes recognise, and otherwise leaves the
    value intact so nothing is silently renamed — the classifier's shared default
    then applies identically on both sides.
    """
    normalised = (value or "").strip().lower()
    return normalised or "strength"

File: apps/web/src/lib/workoutCategories.ts
import {
  categoryForWorkoutType,
  isKnownWorkoutType,
  WORKOUT_TYPE_CATEGORY,
  WORKOUT_TYPES,
} from '@coach/shared';
import type { DayCategory, WorkoutType } from '@coach/shared';

import type { DailyLoopData } from '@/hooks/useDailyLoop';

// Batch 253 (CR236-05): the vocabulary and the classifier now live in
// `@coach/shared`, mirrored entry-for-entry by `services/workout_categories`.
// This module re-exports them so its existing importers do not move, and keeps
// only what is genuinely app-side: the labels and the day-card assembly.
export { categoryForWorkoutType, isKnownWorkoutType, WORKOUT_TYPE_CATEGORY, WORKOUT_TYPES };
export type { DayCategory, WorkoutType };

const LABELS: Record<Exclude<DayCategory, 'rest'>, string> = {
  cycle: 'Cycle',
  weights: 'Weights',
  flexibility: 'Flexibility',
  walk: 'Walk',
};

export function isBikeWorkoutType(workoutType: string | null | undefined): boolean {
  return categoryForWorkoutType(workoutType) === 'cycle';
}

// Clean, human label for a workout-type enum, used in the day-card subtitle.
// Replaces the old per-page `type.replace(/[_-]+/g, ' ')` which leaked the raw
// discipline prefix ("Bike sweet spot", "Bike z2") beneath an already-friendly
// title. The category badge/icon already conveys the discipline, so the label
// drops the `bike_`/`strength_` prefix and reads as the session's character.
const WORKOUT_TYPE_LABELS: Record<string, string> = {
  bike_z2: 'Zone 2',
  bike_endurance: 'Endurance',
  bike_tempo: 'Tempo',
  bike_sweet_spot: 'Sweet spot',
  bike_threshold: 'Threshold',
  bike_vo2: 'VO₂',
  bike_recovery: 'Recovery ride',
  strength: 'Strength',
  mobility: 'Mobility',
  flexibility: 'Mobility',
  walk: 'Walk',
  deliberate_walk: 'Walk',
};

export function workoutTypeLabel(workoutType: string | null | undefined): string {
  const value = (workoutType ?? '').toLowerCase().trim();
  if (!value) return 'Session';
  const mapped = WORKOUT_TYPE_LABELS[value];
  if (mapped) return mapped;
  // Fallback: strip a known discipline prefix, de-underscore, sentence-case.
  const stripped = value.replace(/^(bike|strength|cycle|ride)_/, '');
  const cleaned = stripped.replace(/[_-]+/g, ' ').trim();
  return cleaned.charAt(0).toUpperCase() + cleaned.slice(1);
}

export function dayStateForWorkouts(workouts: Array<Pick<DailyLoopData['plannedWorkouts'][number], 'workoutType'>>): {
  categories: DayCategory[];
  label: string;
  isRest: boolean;
} {
  const categories: Exclude<DayCategory, 'rest'>[] = [];
  for (const workout of workouts) {
    const category = categoryForWorkoutType(workout.workoutType);
    if (!categories.includes(category)) categories.push(category);
  }
  if (categories.length === 0) {
    return { categories: ['rest'], label: 'Rest', isRest: true };
  }
  return {
    categories,
    label: categories.map((category) => LABELS[category]).join(' + '),
    isRest: false,
  };
}
